### PR TITLE
Feature/forex for linkex

### PR DESCRIPTION
**Pull request by j16r**

### CODE REVIEW

File: solidity/contracts/LinkEx.sol [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    if (isFutureBlock()) {
+      return;
+    }
+    historicRate = rate;
+    rate = _rate;
+    rateHeight = block.number;
+  }
+
+  function isFutureBlock() internal view returns (bool) {
+    return rateHeight != 0 && block.number > rateHeight;
```
Comment (se3000) [inline solidity/contracts/LinkEx.sol]: Do we need the `!= 0` check?

File: solidity/contracts/LinkEx.sol [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  function currentRate() public view returns (uint256) {
+    if (isFutureBlock()) {
+      return rate;
+    }
+    return historicRate;
+  }
+
+  function update(uint256 _rate) public {
+    if (isFutureBlock()) {
```
Comment (se3000) [inline solidity/contracts/LinkEx.sol]: What is the scenario that this check is guarding against? I think we could cut it.

**Comment by j16r** [comment]
checking that an update has already been made, so a rate for a block can’t
be changed after the fact

**Comment by se3000** [comment]
I think that could very easily happen with network congestion. Imagine that the gas price is too high for the updates to get in, and so once the minimum gas price drops to the level of the updating transactions, all of the transactions are confirmed in one block. Wouldn't we want to keep the latest value and not throw it out?

**Comment by thodges-gh** [comment]
Maybe the story doesn't capture this scenario very well, since it's written from the perspective of Carol. The rate should be able to be updated multiple times in a single block, but if a requester calls `currentRate()` within the same block which has received an update, they should receive the previous rate. This would mean that when updating the rate, you would still need to check if the rate has been updated in the same block, but instead of failing, you would just not update the previous rate (since it's from a previous block).

File: examples/echo_server/migrations/5_run_log.js [hunk 3]
```diff
@@ -14,7 +13,7 @@ let job = {
   tasks: [{ type: 'HttpPost', params: { url: 'http://localhost:6690' } }]
 }
 
-module.exports = clmigration(async function(truffleDeployer) {
+module.exports = async function(truffleDeployer) {
```
Comment (se3000) [inline examples/echo_server/migrations/5_run_log.js]: 👍

File: solidity/test/TestLinkEx.sol [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+  function testUpdateSameTransaction() public {
+    LinkEx ex = new LinkEx();
+
+    ex.update(8616460799);
+    Assert.equal(ex.currentRate(), 0, "Contract should only return the rate set for future blocks");
+  }
+
+  function testUpdateNewTransaction() public {
+    LinkEx ex = LinkEx(DeployedAddresses.LinkEx());
```
Comment (se3000) [inline solidity/test/TestLinkEx.sol]: 👍

**Pull request merged**